aiosmtpd/controller.py

```diff
@@ -137,39 +144,31 @@ def _run(self, ready_event):
         self.loop.close()
         self.server = None
 
-    def _testconn(self):
-        """
-        Opens a socket connection to the newly launched server, wrapping in an SSL
-        Context if necessary, and read some data from it to ensure that factory()
-        gets invoked.
-        """
-        with ExitStack() as stk:
-            s = stk.enter_context(create_connection((self.hostname, self.port), 1.0))
-            if self.ssl_context:
-                s = stk.enter_context(self.ssl_context.wrap_socket(s))
-            _ = s.recv(1024)
-
     def start(self):
         assert self._thread is None, "SMTP daemon already running"
         ready_event = threading.Event()
+        self._factory_invoked = threading.Event()
         self._thread = threading.Thread(target=self._run, args=(ready_event,))
         self._thread.daemon = True
         self._thread.start()
         # Wait a while until the server is responding.
         ready_event.wait(self.ready_timeout)
         if self._thread_exception is not None:  # pragma: on-wsl
             # See comment about WSL1.0 in the _run() method
-            assert self._thread is not None  # Stupid LGTM.com; see github/codeql#4918
             raise self._thread_exception
+        if not ready_event.is_set():
+            raise TimeoutError("SMTP server failed to start within allotted time")
         # Apparently create_server invokes factory() "lazily", so exceptions in
         # factory() go undetected. To trigger factory() invocation we need to open
         # a connection to the server and 'exchange' some traffic.
         try:
-            self._testconn()
+            self._trigger_server()
         except Exception:
-            # We totally don't care of exceptions experienced by _testconn,
+            # We totally don't care of exceptions experienced by _trigger_server,
             # which _will_ happen if factory() experienced problems.
             pass
+        if not self._factory_invoked.wait(self.ready_timeout):
+            raise TimeoutError("SMTP server not responding within allotted time")
         if self._thread_exception is not None:
             raise self._thread_exception
```

Comment:
If for some reasons `factory()` or `_factory_invoker()` take their sweet time in doing their job, we might miss a "late exception" because when main thread progresses to `if self._thread_exception` line there, `_factory_invoker()` might not have the opportunity yet to set the `self._thread_exception` variable.

Ideally, the timeout should be "what's left after waiting for create_server()", but that adds more complexity and I'm reluctant to do that.